SI-8410 Don't warn fatally on disabled flag

som-snytt opened the pull request:
Since boolean settings can now be set false by user,
summary warnings should not be issued when the flag
is explicitly off (as opposed to unset, default).

In particular, `-Xfatal-warnings` should not fail
if there were no warnings otherwise.

```
$ ~/scala-2.11.2/bin/scalac -d /tmp -deprecation:false test/files/pos/t8410.scala
$ ~/scala-2.11.2/bin/scalac -d /tmp -deprecation:false -Xfatal-warnings test/files/pos/t8410.scala
warning: there were three deprecation warnings; re-run with -deprecation for details
error: No warnings can be incurred under -Xfatal-warnings.
one warning found
one error found

```

After this commit:

```
$ skalac -d /tmp -Xfatal-warnings test/files/pos/t8410.scala
warning: there were three deprecation warnings; re-run with -deprecation for details
error: No warnings can be incurred under -Xfatal-warnings.
one warning found
one error found
$ skalac -d /tmp -deprecation:false -Xfatal-warnings test/files/pos/t8410.scala
```

Similarly for other collecting flags:

```
$ skalac -d /tmp -optimise -Yinline-warnings -deprecation:false -Xfatal-warnings test/files/pos/t8410.scala
test/files/pos/t8410.scala:14: warning: Could not inline required method dropWhile because access level required by callee not matched by caller.
  def k = List(0).dropWhile(_ < 1)   // inlining warns doubly
                           ^
test/files/pos/t8410.scala:14: warning: At the end of the day, could not inline @inline-marked method dropWhile
  def k = List(0).dropWhile(_ < 1)   // inlining warns doubly
                           ^
error: No warnings can be incurred under -Xfatal-warnings.
two warnings found
one error found
$ skalac -d /tmp -optimise -Yinline-warnings:false -deprecation:false -Xfatal-warnings test/files/pos/t8410.scala

```

Footnote: handling of deprecated locals also changed in 2014:

```
$ ~/scala-2.11.0-M7/bin/scalac -d /tmp -deprecation -Xfatal-warnings test/files/pos/t8410.scala
test/files/pos/t8410.scala:8: warning: method f in object Test is deprecated:
  Console println f  // warns
                  ^
error: No warnings can be incurred under -Xfatal-warnings.
one warning found
one error found
$ ~/scala-2.11.0-M8/bin/scalac -d /tmp -deprecation -Xfatal-warnings test/files/pos/t8410.scala
test/files/pos/t8410.scala:5: warning: method _f is deprecated:
  def g = { @deprecated("","") def _f = f ; _f }                   // warns in 2.11.0-M8
                                            ^
test/files/pos/t8410.scala:6: warning: class X is deprecated:
  def x = { @deprecated("","") class X { def x = f } ; new X().x } // warns in 2.11.0-M8
                                                           ^
test/files/pos/t8410.scala:8: warning: method f in object Test is deprecated:
  Console println f  // warns
                  ^
error: No warnings can be incurred under -Xfatal-warnings.
three warnings found
one error found

```

som-snytt committed:
SI-8410 Don't warn fatally on disabled flag

Since boolean settings can now be set false by user,
summary warnings should not be issued when the flag
is explicitly off (as opposed to unset, default).

In particular, `-Xfatal-warnings` should not fail
if there were no warnings otherwise.

```
$ ~/scala-2.11.2/bin/scalac -d /tmp -deprecation:false test/files/pos/t8410.scala
$ ~/scala-2.11.2/bin/scalac -d /tmp -deprecation:false -Xfatal-warnings test/files/pos/t8410.scala
warning: there were three deprecation warnings; re-run with -deprecation for details
error: No warnings can be incurred under -Xfatal-warnings.
one warning found
one error found

```
After this commit:
```
$ skalac -d /tmp -Xfatal-warnings test/files/pos/t8410.scala
warning: there were three deprecation warnings; re-run with -deprecation for details
error: No warnings can be incurred under -Xfatal-warnings.
one warning found
one error found
$ skalac -d /tmp -deprecation:false -Xfatal-warnings test/files/pos/t8410.scala
```
Similarly for other collecting flags:
```
$ skalac -d /tmp -optimise -Yinline-warnings -deprecation:false -Xfatal-warnings test/files/pos/t8410.scala
test/files/pos/t8410.scala:14: warning: Could not inline required method dropWhile because access level required by callee not matched by caller.
  def k = List(0).dropWhile(_ < 1)   // inlining warns doubly
                           ^
test/files/pos/t8410.scala:14: warning: At the end of the day, could not inline @inline-marked method dropWhile
  def k = List(0).dropWhile(_ < 1)   // inlining warns doubly
                           ^
error: No warnings can be incurred under -Xfatal-warnings.
two warnings found
one error found
$ skalac -d /tmp -optimise -Yinline-warnings:false -deprecation:false -Xfatal-warnings test/files/pos/t8410.scala

```
Footnote: handling of deprecated locals also changed in 2014:
```
$ ~/scala-2.11.0-M7/bin/scalac -d /tmp -deprecation -Xfatal-warnings test/files/pos/t8410.scala
test/files/pos/t8410.scala:8: warning: method f in object Test is deprecated:
  Console println f  // warns
                  ^
error: No warnings can be incurred under -Xfatal-warnings.
one warning found
one error found
$ ~/scala-2.11.0-M8/bin/scalac -d /tmp -deprecation -Xfatal-warnings test/files/pos/t8410.scala
test/files/pos/t8410.scala:5: warning: method _f is deprecated:
  def g = { @deprecated("","") def _f = f ; _f }                   // warns in 2.11.0-M8
                                            ^
test/files/pos/t8410.scala:6: warning: class X is deprecated:
  def x = { @deprecated("","") class X { def x = f } ; new X().x } // warns in 2.11.0-M8
                                                           ^
test/files/pos/t8410.scala:8: warning: method f in object Test is deprecated:
  Console println f  // warns
                  ^
error: No warnings can be incurred under -Xfatal-warnings.
three warnings found
one error found

```

gkossakowski commented on src/compiler/scala/tools/nsc/Reporting.scala:
```diff
@@ -32,14 +32,17 @@ trait Reporting extends scala.reflect.internal.Reporting { self: ast.Positions w
-        if (warnings.nonEmpty && (option.isDefault || settings.fatalWarnings)) {
+      def summarize() = {
+        def turnedOff = option.isSetByUser && !option
+        def moreInfos = option.isDefault || settings.fatalWarnings
```
What more infos mean in this context?

som-snytt replied:
I struggled to find a word with the same length as `turnedOff`.  They do want more info about warnings. Nothing to do with symbol infos.  The expr used to mean: either they didn't set the flag (so warnings weren't reported yet) or fatal is on.  I retained the expr because (only in theory) maybe the flag is switched on/off programmatically; so maybe warnings were collected even though flag is now on, in which case fatalWarnings would summarize.

gkossakowski replied:
Why `fatalWarnings` is relevant to whether summarize warnings or not? I know that logic was already there but maybe you could shed some light on this. I don't follow this logic and `git blame` is not very helpful.

som-snytt replied:
I agree.  Another try at clarity: `warn` either reports immediately or buffers; `summarize` says tell me anything you haven't told me yet. No, wait.

OK, it seems to be the product of a massive merge in Feb 2013:

https://github.com/scala/scala/commit/22d315d61b11e95c3a18e1285ca2131f614e13fb#diff-d686e8e67add4d42682f22193d20494b

Delta in the neighborhood:

https://github.com/scala/scala/commit/e0068b908517768e900a3945e483e9c379d728d8#diff-d686e8e67add4d42682f22193d20494b

I don't see anything on 2.10.x.  My assumption was that someone was putting something in the warnings map at some point and needed them flushed, but that was just a guess.  Everything has to go through the reporting front door now.  Unless @adriaanm says something different when he rises from the dead next Halloween.

gkossakowski replied:
The merge you mentioned looks off to me. I analyzed both parents of the merge commit and none of them have changes in that region of the code. My hunch is that `fatalWarnings` has no use here. @som-snytt, could you add another commit that removes call to `fatalWarnings`? I prefer to have it as separate commit so we can revert easily in case of regression.

Also, I think vals are fine above, we don't need defs, right?

som-snytt replied:
Sure, that would be my conclusion, too.  Red Herring sandwich.

I use vals or defs depending on whether I'm thinking Mos Def or Prince Valiant (aka Prince Val).

som-snytt committed:
SI-8410 Summarize if warnings and not disabled

Simplify the conditional and emit the summary if
there were buffered warnings and the user didn't
explicitly disable this setting.

The working assumption is that using fatal-warnings
to turn on summaries was either an outdated heuristic
or just due to a faulty merge. Or wait, was it for
-feature, which is not enabled when warnings are
issued against it?

som-snytt commented:
Idea was to nail down condition for issuing summary.  Used to be: flag is default (and therefore false) or fatal-warnings.  Now is just: flag is default or set true, i.e., not explicitly turned off. That covers the case of issuing warnings with the flag off, then enabling the flag.  (Maybe that never happens. Maybe there are corner cases in the way -feature works, etc.)

gkossakowski commented:
LGTM

Excellent work!

puffnfresh commented:
:fireworks:

Thank you!
